### PR TITLE
Class estimator

### PR DESCRIPTION
**Pull request by johanmazoyer**
In practice changes are transparent for the correction, i did not touche the optical side, just the python organization 

Big changes:
I finally defined my concatenation function for Optical System. We can now change the system completely transparently for the code, this is ready for Roman ! 
Initializaiton of the estimation class.
The Example_run_asterix is now a big test file that you can run when you want to push to master. You can comment part of it to go faster but you nedd to run it fully before push ! 
The tutorial is ready, you can have a look !

### CODE REVIEW

File: Asterix/tuto_asterix_model.py [hunk 1]
```diff
@@ -44,79 +44,225 @@
...
+DMconfig.update({'DM1_z_position': -15e-2})  # meter
+DMconfig.update({'DM1_active': True})
+DMnew = OptSy.deformable_mirror(modelconfig,
+                                DMconfig,
+                                load_fits=False,
+                                save_fits=True,
+                                Name_DM='DM1',
+                                model_dir=model_dir,
+                                Model_local_dir=Model_local_dir)
+# the Name_DM in this function is to be understand as the type of DM you want to use (DM3 is a BMC32x32 type DM
```
Comment (rgalicher) [inline Asterix/tuto_asterix_model.py:249]: Hum. On devrait changer ça (peut-être DMpup et DMoutpup?). Que DM3 soit un DM1, c'est bizarre. :-)

File: Asterix/Main_EFC_THD.py [hunk 2]
```diff
@@ -115,18 +115,39 @@ def create_interaction_matrices(parameter_file,
...
+
+    DM3 = OptSy.deformable_mirror(modelconfig,
+                                DMconfig,
+                                load_fits=not DM3_creating_pushact,
+                                save_fits=DM3_creating_pushact,
+                                Name_DM='DM3',
+                                model_dir=model_dir,
+                                Model_local_dir=Model_local_dir)
+
+    # we also need to "clear" the apod plane because the THD2 is like that
```
Comment (rgalicher) [inline Asterix/Main_EFC_THD.py:141]: Same remark as elsewhere "because there is no such plane on the THD2 bench"

File: Asterix/Main_EFC_THD.py [hunk 3]
```diff
@@ -475,12 +497,42 @@ def correctionLoop(parameter_file,
...
+                                model_dir=model_dir,
+                                Model_local_dir=Model_local_dir)
+
+    DM3 = OptSy.deformable_mirror(modelconfig,
+                                DMconfig,
+                                Name_DM='DM3',
+                                model_dir=model_dir,
+                                Model_local_dir=Model_local_dir)
+
+    # we also need to "clear" the apod plane because the THD2 is like that
```
Comment (rgalicher) [inline Asterix/Main_EFC_THD.py:513]: "because there is no such plane on the THD2 bench"

File: Asterix/Main_EFC_THD.py [hunk 4]
```diff
@@ -475,12 +497,42 @@ def correctionLoop(parameter_file,
...
+    Coronaconfig.update({'filename_instr_apod': "ClearPlane"})
+    corono = OptSy.coronagraph(modelconfig, Coronaconfig, model_dir=model_dir)
+    # and then just concatenate
+    thd2 = OptSy.Testbed([pup_round, DM1, DM3, corono],
+                                ["entrancepupil", "DM1", "DM3", "corono"])
+
+    # TODO Because the code is currently setup heavily on the
```
Comment (rgalicher) [inline Asterix/Main_EFC_THD.py:520]: Not sure I get it but ok if the TODO is for you. ;-)

File: Asterix/Main_EFC_THD.py [hunk 5]
```diff
@@ -475,12 +497,42 @@ def correctionLoop(parameter_file,
...
+    # there is an error if it thinks they are not defined (although in practice
+    # they are).
+    thd2.entrancepupil = pup_round
+    thd2.DM1 = DM1
+    thd2.DM3 = DM3
+    thd2.corono = corono
+    # In practice this is done inside the Testbed initialization already !
+    # and these lines are useless and I only put them to calm
+    # To be removed when the correction and estimation class are done
+    # in which case these class will be able when these things exists or not
```
Comment (rgalicher) [inline Asterix/Main_EFC_THD.py:531]: Now the estimation and correction class exist. Can you remove these lines?

File: Asterix/Main_EFC_THD.py [hunk 6]
```diff
@@ -602,36 +654,34 @@ def correctionLoop(parameter_file,
...
 
     amplitude_abb_up = ampfinal
     phase_abb_up = phase_up
 
     ## To convert in photon flux
-    # We can probably replace here by transmission !
-    contrast_to_photons = (np.sum(thd2.entrancepupil.pup) /
-                           np.sum(thd2.corono.lyot_pup.pup) * nb_photons *
-                           thd2.maxPSF / thd2.sumPSF)
+    contrast_to_photons = 1 / thd2.transmission() * nb_photons * thd2.maxPSF / thd2.sumPSF
```
Comment (rgalicher) [inline Asterix/Main_EFC_THD.py:683]: Why maxPSF/sumPSF and why 1/transmission and not transmission?
transmission  = sum(pupil out) / sum(pupil in)
nb photon out = nb photons * transmission

Not sure what contrast_to_photons represents.

File: Asterix/Optical_System_functions.py [hunk 7]
```diff
@@ -283,24 +297,26 @@ def todetector_Intensity(self,
...
         """
         clear_entrance_pupil = phase_ampl.roundpupil(self.dim_overpad_pupil,
                                                      self.prad)
 
-        exit_EF = self.EF_through(**kwargs)
+        # all parameter can be passed here, but in the case there is a coronagraph,
+        # we pass noFPM = True and noentrance Field by default
+        exit_EF = self.EF_through(entrance_EF=1., noFPM=noFPM, **kwargs)
 
         throughput_loss = np.sum(
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:321]: Pourquoi throughput_loss? Ce n'est pas juste  throughput?

File: Asterix/Optical_System_functions.py [hunk 8]
```diff
@@ -313,36 +329,46 @@ def EF_from_phase_and_ampl(self,
...
         if np.iscomplexobj(phase_abb) or np.iscomplexobj(ampl_abb):
             raise Exception(
                 "phase_abb and ampl_abb must be real arrays or float, not complex"
             )
 
-        if (phase_abb == 0.).all() and (ampl_abb == 0).all():
+        if isinstance(phase_abb, (int, float, np.float)):
+            phase_abb = np.full(
+                (self.dim_overpad_pupil, self.dim_overpad_pupil),
+                np.float(phase_abb))
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:361]: What does this do?

File: Asterix/Optical_System_functions.py [hunk 9]
```diff
@@ -829,10 +878,10 @@ def EF_through(self,
...
-
-                name_plane = 'FPM' + '_wl{}'.format(int(wavelength * 1e9))
-                useful.save_plane_in_fits(dir_save_all_planes, name_plane,
-                                          FPmsk)
+                if not noFPM:
+                    name_plane = 'FPM' + '_wl{}'.format(int(wavelength * 1e9))
+                    useful.save_plane_in_fits(dir_save_all_planes, name_plane,
+                                              FPmsk)
 
                 name_plane = 'EF_FP_after_FPM' + '_wl{}'.format(
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:886]: name_plane écrase celui défini dans "if not noFPM" deux lignes plus haut non ?

File: Asterix/Optical_System_functions.py [hunk 10]
```diff
@@ -897,7 +946,6 @@ def FQPM(self):
             maxdimension_array_fpm = np.max(self.dim_fp_fft)
         else:
             maxdimension_array_fpm = self.dim_im
-            print("you should really not simulate FQPM with MFT")
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:900]: Ah, j'ai gagné ;-)

File: Asterix/Optical_System_functions.py [hunk 11]
```diff
@@ -924,7 +972,6 @@ def FQPM(self):
                 dim_fp = self.dim_fp_fft[i]
             else:
                 dim_fp = self.dim_im
-                print("really you should not do that")
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:927]: J'ai vraiment gagné. :-D

File: Asterix/Optical_System_functions.py [hunk 12]
```diff
@@ -1387,27 +1434,29 @@ def creatingpushact_inpup(self,
...
+        ) and (os.path.exists(Model_local_dir + Name_pushact_inpup_fits +
+                              '_inPup_imag.fits')):
             DM_pushact_inpup_real = fits.getdata(
                 os.path.join(Model_local_dir,
-                             Name_pushact_fits + '_inPup_real.fits'))
+                             Name_pushact_inpup_fits + '_inPup_real.fits'))
             DM_pushact_inpup_imag = fits.getdata(
                 os.path.join(Model_local_dir,
-                             Name_pushact_fits + '_inPup_imag.fits'))
+                             Name_pushact_inpup_fits + '_inPup_imag.fits'))
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:1459]: Comment about the question that is below "Do we need a pupil?"
I believe we do for the Fresnel propagation. Need to think about it.

File: Asterix/Optical_System_functions.py [hunk 13]
```diff
@@ -1595,163 +1646,218 @@ def voltage_to_phase(self, actu_vect, wavelength=None):
...
-        dir_save_all_planes : default None. directory to save all plane in fits if save_all_planes_to_fits = True
-                    
+        # we store the name of all the sub systems
+        self.subsystems = list_os_names
 
-        Returns
+        # we concatenate the Optical Element starting by the end
+        for num_optical_sys in range(len(list_os)):
+
+            # we firt check that all variables in the list are optical systems
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:1710]: first

File: Asterix/Optical_System_functions.py [hunk 14]
```diff
@@ -1595,163 +1646,218 @@ def voltage_to_phase(self, actu_vect, wavelength=None):
...
+                raise Exception("list_os[" + str(num_optical_sys) +
+                                "] is not an optical system")
+
+            if list_os[num_optical_sys].modelconfig != self.modelconfig:
+                print("")
+                raise Exception(
+                    "All optical systems need to be defined with the same initial modelconfig!"
+                )
+
+            # if the os if a DM we increase the number of DM counter and
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:1722]: if the os is a DM

File: Asterix/Optical_System_functions.py [hunk 15]
```diff
@@ -1595,163 +1646,218 @@ def voltage_to_phase(self, actu_vect, wavelength=None):
...
+
+            # concatenation of the EF_through functions
+            self.EF_through = _concat_fun(list_os[num_optical_sys].EF_through,
+                                          self.EF_through)
+
+            # we add all systems to the Optical System so that they can be accessed
+            vars(self)[
+                list_os_names[num_optical_sys]] = list_os[num_optical_sys]
+
+        # we remove doublons
```
Comment (rgalicher) [inline Asterix/Optical_System_functions.py:1759]: doubloons

**Review by rgalicher — APPROVED** [review]

File: Asterix/Optical_System_functions.py (hunk 12 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:1459]: yes, this is one of the question we need to think about, I have been thinking about it. this is #1 on my list for next wednesday meeting. I think we do, but there are other problems !

File: Asterix/estimator.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+        Load or save
+
+        AUTHOR : Johan Mazoyer
+        -------------------------------------------------- """
+
+        if isinstance(testbed, OptSy.Optical_System) == False:
+            raise Exception("testbed must be an Optical_System objet")
+
+        self.is_focal_plane = None
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/estimator.py:61]: done iun later version

File: Asterix/tuto_asterix_model.py [hunk 17]
```diff
@@ -44,79 +44,225 @@
...
-
+# by defaulf all pupil have the diameter of the telescope but smaller pupil / larger can be define
+# This can be useful for Lyot stop
 # Clear pupil of radius equal to 100 pixel
-pup_round_100 = instr.pupil(modelconfig, prad=100)
-
-# these are Opitcal System objects, so we can easily access their PSF for example:
-psf_roman = pup_roman.todetector_Intensity()
-useful.quickfits(psf_roman, dir=result_dir)
+pup_round_100 = OptSy.pupil(modelconfig, prad=100)
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:63]: nowhere, this is just a tutorial to show how you can do it

File: Asterix/Optical_System_functions.py (hunk 11 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:927]: :-D

File: Asterix/tuto_asterix_model.py [hunk 18]
```diff
@@ -44,79 +44,225 @@
...
+# TODO a better way to define this pupil can be to indicate the radius as a percentage of
+# the radius of the initial pupil
 
+# # Careful, pup_roman is not an array, it is an Optical System object.
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:67]: pas compris.This is a tutorial, I'm not using the object, just showing them

File: Asterix/tuto_asterix_model.py [hunk 19]
```diff
@@ -44,79 +44,225 @@
...
+                              Name_DM='DM3',
+                              model_dir=model_dir,
+                              Model_local_dir=Model_local_dir)
+
+# DMs are also optical systems but have another parameter to control them, DMphase
+# measure the EF throught the DM
+EF_though_DM = DM3.EF_through(entrance_EF=1., DMphase=0.)
+# plus all normal function of optical systems (todetector, todetector_Intensity)
+
+# for example, something like can be funny:
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:167]: electrical field just went "though DM3", which in this case is a multiplincation by exp(i phi pahse)

File: Asterix/tuto_asterix_model.py [hunk 20]
```diff
@@ -44,79 +44,225 @@
...
+EF_though_DM = DM3.EF_through(entrance_EF=aberrated_EF, DMphase=phase)
+
+# Now that we have all these Optical Systems defined, we can play with it and concatenate them.
+# The concatenate function takes 2 parameters:
+#                               - a list of Optical Systems
+#                               - A list of the same size of the name of those system so that you can access it
+testbed_1DM = OptSy.Testbed([pup_round, DM3, corono],
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:174]: yes, i should write that

File: Asterix/tuto_asterix_model.py [hunk 21]
```diff
@@ -44,79 +44,225 @@
...
+print("number of DMs in testbed_1DM_romanpup:",
+      testbed_1DM_romanpup.number_DMs)
+print("name of the DMs: ", testbed_1DM_romanpup.name_of_DMs)
+print("number of actuators in each DMs: ",
+      testbed_1DM_romanpup.number_of_acts_in_DMs)
+
+# if we want to define exactly the thd2, we need to add a second DM off pupil plane.
+# This can take som time to initialize becasue the DM is off pupil.
+
+# We need to increase the number of pixel in the pupil if we add another DM. I'll put it at the minimum to go faster
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:201]: yes, we need to talk about that I think your chaecking parameter on minimum pixel is not enough ! 
dimpup = 220 is running, but gives clearly wring results

File: Asterix/tuto_asterix_model.py (hunk 1 above)
Comment (johanmazoyer) [reply Asterix/tuto_asterix_model.py:249]: oui c'est sur ma liste un jour, mais pas extrement prioritaire vu qu'on va rester dans cette config un moment

File: Asterix/Main_EFC_THD.py (hunk 4 above)
Comment (johanmazoyer) [reply Asterix/Main_EFC_THD.py:520]: has been removed

File: Asterix/Main_EFC_THD.py (hunk 5 above)
Comment (johanmazoyer) [reply Asterix/Main_EFC_THD.py:531]: almost !

File: Asterix/Optical_System_functions.py (hunk 7 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:321]: done

File: Asterix/Main_EFC_THD.py (hunk 6 above)
Comment (johanmazoyer) [reply Asterix/Main_EFC_THD.py:683]: this has been changed

File: Asterix/Optical_System_functions.py (hunk 8 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:361]: if you put phase_abb = 0. or  phase_abb = 0 it transform to np.zeros(self.dim_overpad_pupil, self.dim_overpad_pupil)

also works for any constant (e.g. phase_abb = np.pi)

File: Asterix/Optical_System_functions.py (hunk 9 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:886]: yes but i saved the thing I want in between

File: Asterix/Optical_System_functions.py (hunk 10 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:900]: ahah oui !

File: Asterix/tuto_asterix_model.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rgalicher) [reply Asterix/tuto_asterix_model.py:167]: Why "can be funny"?

File: Asterix/Optical_System_functions.py (hunk 8 above)
Comment (rgalicher) [reply Asterix/Optical_System_functions.py:361]: ok but do we really need this? exp(i phase) multiplies a 2D-array. If phase_abb = Cte then, pup = pup0*exp(i phase_abb) works. No need to create and 2D-array with constant value for all pixels (will same memory).

**Comment by rgalicher** [comment]
Ok. Let's go then!

File: Asterix/Optical_System_functions.py (hunk 8 above)
Comment (johanmazoyer) [reply Asterix/Optical_System_functions.py:361]: Yes, you're perfectly right, we can easily save time! this will be done today

**Pull request merged**